### PR TITLE
Do validation on emailaddress and not NameId

### PR DESCRIPTION
**Pull request by pablothedude**
The validation of a succesful response should be done
on the email attribute and not the NameID of the response.

### CODE REVIEW

File: src/Surfnet/AzureMfa/Application/Service/AzureMfaService.php [hunk 1]
```diff
@@ -38,6 +38,8 @@
  */
 class AzureMfaService
 {
+    const SAML_EMAIL_ATTRIBUTE = 'urn:mace:dir:attribute-def:emailAddress';
```
Comment (thijskh) [inline src/Surfnet/AzureMfa/Application/Service/AzureMfaService.php:41]: That does not exist in mace-dir. The emailadresss is in `...:mail`. See: https://www.internet2.edu/products-services/trust-identity/mace-registries/urnmace-namespace/urn-mace-dir-registry/urn-mace-dir-attribute-def/

File: src/Surfnet/AzureMfa/Application/Service/AzureMfaService.php [hunk 2]
```diff
@@ -194,9 +196,18 @@ public function handleResponse(Request $request): User
...
+            throw new InvalidMfaNameIdException(
+                'The email attribute from the Azure MFA assertion was missing'
+            );
+        }
+        $emailAddressAttribute = $attributes[self::SAML_EMAIL_ATTRIBUTE][0];
+
+        if ($emailAddressAttribute !== $user->getEmailAddress()->getEmailAddress()) {
             throw new InvalidMfaNameIdException(
-                'The NameId from the Azure MFA assertion did not match the NameId provided during registration'
+                'The email attribute from the Azure MFA assertion did not match the NameId provided during registration'
```
[the remark this answers is not in the document]
Comment (pablothedude) [reply src/Surfnet/AzureMfa/Application/Service/AzureMfaService.php:210]: Good suggestion, will change that accordingly!

File: src/Surfnet/AzureMfa/Application/Service/AzureMfaService.php [hunk 3]
```diff
@@ -194,9 +196,18 @@ public function handleResponse(Request $request): User
...
+        $attributes = $assertion->getAttributes();
+
+        if (!isset($attributes[self::SAML_EMAIL_ATTRIBUTE]) || !isset($attributes[self::SAML_EMAIL_ATTRIBUTE][0])) {
+            throw new InvalidMfaNameIdException(
+                'The email attribute from the Azure MFA assertion was missing'
+            );
+        }
+        $emailAddressAttribute = $attributes[self::SAML_EMAIL_ATTRIBUTE][0];
+
+        if ($emailAddressAttribute !== $user->getEmailAddress()->getEmailAddress()) {
```
[the remark this answers is not in the document]
Comment (pablothedude) [reply src/Surfnet/AzureMfa/Application/Service/AzureMfaService.php:208]: I thought initially that the UPN would be the NameID, or at least that you could configure your 
email address as NameID. So frankly I'm no domain expert regarding AzureMFA or how you could configure how to release the UPN.

https://docs.microsoft.com/nl-nl/azure/active-directory/develop/active-directory-saml-claims-customization
https://www.pivotaltracker.com/story/show/169312724

Maybe an unverified email address could be released or that it has to be validated in combination with another attribute to see if so?

File: dev/Mock/MockGateway.php [hunk 4]
```diff
@@ -69,6 +69,7 @@ public function __construct(
      * @param Request $request
      * @param string $fullRequestUri
      * @return Response
+     * @throws \Exception
```
Comment (MKodde) [inline dev/Mock/MockGateway.php:72]: NIt: if possible, please import the `\Exception` class

File: dev/Mock/MockGateway.php [hunk 5]
```diff
@@ -308,6 +313,9 @@ private function createNewAssertion($nameId, $authnContextClassRef, $destination
             'Format' => Constants::NAMEID_UNSPECIFIED,
             'Value' => $nameId,
         ]);
+        $newAssertion->setAttributes([
+            'urn:mace:dir:attribute-def:emailAddress' => [$emailAddress]
```
Comment (MKodde) [inline dev/Mock/MockGateway.php:317]: As Thijs pointed out, this urn:mace attribute is not in the dictionary. According to the page linked it should be; `urn:mace:dir:attribute-def:mail` This was also incorrectly fixed in `AzureMfaService.php`. There it now reads `email`

File: src/Surfnet/AzureMfa/Application/Service/AzureMfaService.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (MKodde) [reply src/Surfnet/AzureMfa/Application/Service/AzureMfaService.php:210]: Nit: ~email-address~  email address (remove the dash, the English do not use the koppelstreepje like we do :) )

File: templates/dev/mock-acs.html.twig [hunk 6]
```diff
@@ -3,7 +3,9 @@
     <b>Note: </b> javascript is disabled, please click the button below to proceed
 </noscript>
 One moment please...
-
+<br/>
+TODO: Add emailaddress field to use instead of NameID
```
Comment (MKodde) [inline templates/dev/mock-acs.html.twig:7]: Is this work in progress or work for some other time? Maybe open a ticket if this should be addressed at a later point.

File: src/Surfnet/AzureMfa/Application/Service/AzureMfaService.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (phavekes) [reply src/Surfnet/AzureMfa/Application/Service/AzureMfaService.php:208]: Yes, the nameid can be configured to return the email-address. And the upn should be the same as the email address. But both don't **need** to be true.
By comparing the email address provides by the user to the email address(es) returned in the IDP response, the behaviour is more logical, and debugging is easier.

File: dev/Mock/MockGateway.php (hunk 5 above)
Comment (pablothedude) [reply dev/Mock/MockGateway.php]: Changed it to be up to the user, now a custom set of attributes could be set win the MockIdP  so this is also better testable.

File: templates/dev/mock-acs.html.twig (hunk 6 above)
Comment (pablothedude) [reply templates/dev/mock-acs.html.twig]: Took some effort to do it, now so it would become better testable and I also added a couple of functional tests.

**Comment by pablothedude** [comment]
@MKodde I've changed the mock IdP so we could release custom attributes to test the flow.
I saw a security warning for NPM but there currently is now fix and it seems to only affect the webpack-dev environment.
https://www.npmjs.com/advisories/1486

**Pull request merged**